handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -578,22 +621,37 @@ private int readPlaintextData(final ByteBuffer dst) throws SSLException {
         } else {
             final int limit = dst.limit();
             final int len = min(maxEncryptedPacketLength0(), limit - pos);
-            final ByteBuf buf = alloc.directBuffer(len);
-            try {
-                sslRead = SSL.readFromSSL(ssl, memoryAddress(buf), len);
+            try (Buffer buf = alloc.allocate(len)) {
+                SslRead read = new SslRead(ssl, len);
+                buf.forEachWritable(0, read);
+                sslRead = read.sslRead;
                 if (sslRead > 0) {
-                    dst.limit(pos + sslRead);
-                    buf.getBytes(buf.readerIndex(), dst);
-                    dst.limit(limit);
+                    buf.copyInto(0, dst, dst.position(), sslRead);
+                    dst.position(dst.position() + sslRead);
                 }
-            } finally {
-                buf.release();
             }
         }
 
         return sslRead;
     }
 
+    static final class SslRead implements WritableComponentProcessor<RuntimeException> {
+        private final long ssl;
+        private final int len;
+        private int sslRead;
+
+        SslRead(long ssl, int len) {
+            this.ssl = ssl;
+            this.len = len;
+        }
+
+        @Override
+        public boolean process(int index, WritableComponent component) {
+            sslRead = SSL.readFromSSL(ssl, component.writableNativeAddress(), len);
```

Comment:
same comment as above related to GC the component

Reply:
Like if you call `forEachWritable` and then immediately leak the buffer?
We can make the `forEach` methods include a reachability fence at the end.

Reply:
I think that should be a separate PR, though.

Reply:
PR: https://github.com/netty/netty/pull/12302